cranelift/codegen/src/opts/bitops.isle

```diff
@@ -104,3 +104,44 @@
 (rule (simplify (sshr ty (bor ty (ineg ty x) x) (iconst ty (u64_from_imm64 shift_amt))))
       (if-let $true (u64_eq shift_amt (u64_sub (ty_bits ty) 1)))
       (bmask ty x))
+
+;; Matches any expressions that preserve "truthiness".
+;; i.e. If the input is zero it remains zero, and if it is nonzero it can have
+;; a different value as long as it is still nonzero.
+(decl pure multi truthy (Value) Value)
+(rule (truthy (sextend _ x)) x)
+(rule (truthy (uextend _ x)) x)
+(rule (truthy (bmask _ x)) x)
+(rule (truthy (ineg _ x)) x)
+(rule (truthy (bswap _ x)) x)
+(rule (truthy (bitrev _ x)) x)
+(rule (truthy (popcnt _ x)) x)
+(rule (truthy (rotl _ x _)) x)
+(rule (truthy (rotr _ x _)) x)
+(rule (truthy (select _ x (iconst _ (u64_from_imm64 (u64_nonzero _))) (iconst _ (u64_from_imm64 0)))) x)
+;; (ne ty (iconst 0) v) also is canonicalized into this form via another rule
+(rule (truthy (ne _ x (iconst _ (u64_from_imm64 0)))) x)
+
+;; All of these expressions don't care about their input as long as it is truthy.
+;; so we can remove expressions that preserve that property from the input.
+(rule (simplify (bmask ty v)) (if-let x (truthy v)) (bmask ty x))
+(rule (simplify (select ty v lhs rhs)) (if-let c (truthy v)) (select ty c lhs rhs))
+;; (ne ty (iconst 0) v) also is canonicalized into this form via another rule
+(rule (simplify (ne cty v zero @ (iconst ty (u64_from_imm64 0))))
+      (if-let c (truthy v))
+      (if-let (value_type ty) c)
+      (ne cty c zero))
```

Comment:
Nice job spotting `(ne v 0)` as also truthy!

I'm having a little trouble following the constraints on this rule. I think what you're doing is that the operands to `icmp` must both have the same type, so if `truthy` unwraps `v` to a different type then the rule doesn't apply. Is that right?

Contrary to what I said before about re-using the existing `iconst`, I think this rule will be better if we reconstruct the constant at whatever type `c` has. In the case where the types do match, we'll go through the GVN map and discover that the new `iconst` is identical to the old one, which is a little silly; but the rule will also succeed when the types don't match, making it more useful.

```suggestion
(rule (simplify (ne cty v (iconst _ (u64_from_imm64 0))))
      (if-let c (truthy v))
      (if-let (value_type ty) c)
      (ne cty c (iconst ty (u64_from_imm64 0))))
```

That said, I believe this rule is also correct the way you wrote it; feel free to merge as-is.

In either case, it would be nice to add a comment, either saying that we're constraining `value_type` to be equal to the type of `zero`, or saying that we're extracting the type of `c` to make sure that the zero has the same type.